### ISSUE TITLE
Unexpected dask dataframe materialization

### ISSUE THREAD

**Issue by adamklein** [dask/distributed#1145]
Please see inline comments

	import dask
	import numpy as np
	import pandas as pd
	import dask.dataframe as dd

	from distributed import Client, LocalCluster

	def create_df(i):
		return pd.DataFrame(np.random.randn(1000, 10),
							index=list(range(i*1000, (i*1000)+1000)),
							columns=[str(i) for i in range(10)])

	def augment_ddf(i, ddf):
		# HUH?? I expect ddf to be a dask dataframe, not a fully materialized pandas dataframe !?
		#       I want to only materialize a slice of it for local computation.
		#       The problem is each augment_ddf gets a full materialized copy of ddf, and this is
		#           blowing up my cluster workers.
		#       Seems like it has to do something with state['cache'] holding a fully materialized version?
		#       How can I force this to be lazy and not move data around??
		slice = ddf.loc[i*1000:i*1000+999]    # .compute() <-- I expect to use this!
		slice = slice.assign(X=np.random.randn(1000))
		return slice

	if __name__ == "__main__":
		# with dask.set_options(get=dask.get):
		with LocalCluster() as cluster:
			with Client(cluster) as client:
				dfs = [dask.delayed(create_df)(i) for i in range(10)]
				ddf = dd.from_delayed(dfs)

				dfs2 = [dask.delayed(augment_df)(i, ddf) for i in range(10)]
				ddf2 = dd.from_delayed(dfs2)

				result = client.persist(ddf2)

				assert(isinstance(result, dask.dataframe.DataFrame))
				assert(len(result) == 10000)

**Comment by mrocklin** [comment]
You can't doubly-delay things.  Delayed functions should expect concrete objects.  Instead you might immediately run a function on your dask dataframe to produce a new dataframe.

```python
dfs2 = [ddf.loc[...] for ...]
ddf2 = dd.concat(dfs2, axis=0)
```

These operations are already lazy, so there is no need to delay them.

**Comment by adamklein** [comment]
Ok, getting closer. But instead of `dfs2=[ddf.loc[...] for ...]`, I basically want `dfs2=[transform(ddf) for ...]` where `transform` is a node that can be run in parallel as part of the DAG execution, and that accepts a dask dataframe that is unmaterialized because I don't know apriori how I will slice it and there's no need to send the whole reified dataframe to the function (the slicing logic lives in `transform`). Is this possible? Seems like a very valuable use case - at least for me :)

**Comment by mrocklin** [comment]
It sounds like you don't know the shape of your DAG before you're halfway done with the computation.  Is this true?  If so then you will probably have to switch to the dask.distributed API.  See http://distributed.readthedocs.io/en/latest/task-launch.html

**Comment by adamklein** [comment]
I guess my confusion is I didn't know dask dataframe itself has to be coupled to DAG execution. I was hoping I could build a dataframe that lives in distributed worker memory and just pass a pointer to it through my DAG, materializing only parts of it as needed (and fully discarding when no longer needed). Maybe I can achieve something similar with client.submit, but I need to think about it more.

**Comment by mrocklin** [comment]
Dask.dataframes are backed by graphs where every node is a single python function running on some data.  Every dask.dataframe operation immediately expands this graph with several nodes.

**Comment by adamklein** [comment]
I think this is basically a misunderstanding of behavior on my part, so not really an issue - I will close. Thanks!!
